### PR TITLE
tools/dm: add doc for cluster operations

### PR DESCRIPTION
**Pull request by dcalvin**
Added Doc for DM Cluster Operations.

Link to be updated:

-  (/tools/dm/practice.md#step-4-start-the-data-synchronization-task)
 - (/tools/dm/manually-handling-sharding-ddl-locks.md#scenario-2-some-dm-workers-restart-during-the-ddl-unlocking-process)

### CODE REVIEW

File: tools/dm/cluster-operations.md [hunk 1]
```diff
@@ -0,0 +1,410 @@
...
+    ```
+
+- 停止 DM-master，然后重启。
+
+    ```bash
+    $ ansible-playbook stop.yml --tags=dm-master
+    $ ansible-playbook start.yml --tags=dm-master
+    ```
+
+### 重启 dmctl
```
Comment (GregoryIan) [inline tools/dm/cluster-operations.md]: dmctl 部分没必要加，因为他的定义就是命令行工具，不是持久运行程序

File: tools/dm/cluster-operations.md [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        $ ansible-playbook -i hosts.ini create_users.yml -u root -k
+        ```
+
+        该步在 `172.16.10.74` 机器上创建了一个 `tidb` 用户，设置了 sudo 规则，并为中控机与该机器配置了 SSH 互信。
+
+2. 修改 `inventory.ini` 文件，创建新 DM-worker 实例 `dm_worker3`。
+
+    ```
+    [dm_worker_servers]
+    dm_worker1 ansible_host=172.16.10.72 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
```
Comment (GregoryIan) [inline tools/dm/cluster-operations.md]: 可以添加一下 source_id

Comment (dcalvin) [reply tools/dm/cluster-operations.md]: “所有”示例具体范围，是指 inventory.ini 文件的示例？另外，source_id的值是？

File: tools/dm/cluster-operations.md [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+**总结**：尽量避免在 sharding DDL 同步过程中重启 DM-worker。
+
+#### DM-master 重启事项
+
+由 DM-master 维护的信息包括以下两种。重启 DM-master 不会持久化保存这些信息的相关数据。
+
+- 任务与 DM-master 之间的对应关系
+- Sharding DDL lock 的相关信息
+
+DM-master 重启时会自动向每个 DM-worker 实例请求任务信息，重建任务与 DM-worker 直接的对应关系，并从每个 DM-worker 实例获取 sharding DDL 信息。这样，就可以准确重建相应的 DDL lock，也可以自动解除 sharding DDL lock。
```
Comment (csuzhangxc) [inline tools/dm/cluster-operations.md]:
```suggestion
DM-master 重启时会自动向每个 DM-worker 实例请求任务信息，重建任务与 DM-worker 之间的对应关系，并从每个 DM-worker 实例获取 sharding DDL 信息。这样，就可以准确重建相应的 DDL lock，也可以自动解除 sharding DDL lock。
```

File: tools/dm/cluster-operations.md [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+
+2. 修改 `inventory.ini` 文件，创建新 DM-worker 实例 `dm_worker3`。
+
+    ```
+    [dm_worker_servers]
+    dm_worker1 source-id: "instance-1" ansible_host=172.16.10.72 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
+
+    dm_worker2  source-id: "instance-2" ansible_host=172.16.10.73 server_id=102 mysql_host=172.16.10.82 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
+
+    dm_worker3 source-id: "instance-3" ansible_host=172.16.10.74 server_id=103 mysql_host=172.16.10.83 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
```
Comment (csuzhangxc) [inline tools/dm/cluster-operations.md]: `source-id: "instance-*"` in these three Lines should be changed  to `source_id="instance-*"`

File: tools/dm/cluster-operations.md [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+
+2. 修改 `inventory.ini` 文件，注释或删除 `dm_worker3` 实例所在行。
+
+    ```
+    [dm_worker_servers]
+    dm_worker1 source-id: "instance-1" ansible_host=172.16.10.72 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
+
+    dm_worker2 source-id: "instance-2" ansible_host=172.16.10.73 server_id=102 mysql_host=172.16.10.82 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
+
+    # dm_worker3 source-id: "instance-3" ansible_host=172.16.10.74 server_id=103 mysql_host=172.16.10.83 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306 # Comment or delete this line
```
Comment (csuzhangxc) [inline tools/dm/cluster-operations.md]: ditto

File: tools/dm/cluster-operations.md [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+3.  修改 `inventory.ini` 文件，为新 DM-worker 实例添加相关信息。
+
+    修改 `inventory.ini` 文件。注释或删除旧 `dm_worker1` 实例所在行；同时为新 `dm_worker1` 实例添加相关信息。
+
+    ```ini
+    [dm_worker_servers]
+    dm_worker1 source-id: "instance-1" ansible_host=172.16.10.75 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
+    # dm_worker1 ansible_host=172.16.10.72 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
+
+    dm_worker2 source-id: "instance-2" ansible_host=172.16.10.73 server_id=102 mysql_host=172.16.10.82 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
```
Comment (csuzhangxc) [inline tools/dm/cluster-operations.md]: ditto

File: tools/dm/cluster-operations.md [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+4. 虚拟 IP 环境下的上游主从实例执行切换。 
+5. 使用 `switch-relay-master` 命令通知 relay 处理单元进行主从切换。
+6. 使用 `resume-relay` 命令恢复 relay 处理，从新 master 实例读取 binlog。
+7. 使用 `resume-task` 命令恢复之前的同步任务。
+
+### 变更 IP 后的主从切换
+
+1. 使用 `query-status` 命令确认 relay 处理单元已获取主从切换前 master 实例的所有 binlog（`relayCatchUpMaster`）。
+2. 使用 `stop-task` 停止所有运行任务。
+3. 修改 DM-worker 配置，并使用 Ansible 对 DM-worker 进行滚动升级操作。
```
Comment (csuzhangxc) [inline tools/dm/cluster-operations.md]:
```suggestion
3. 修改 DM-worker 配置，并使用 DM-Ansible 对 DM-worker 进行滚动升级操作。
```

File: tools/dm/cluster-operations.md [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+4. 更新 `task.yaml` 文件以及 `mysql-instances / config` 配置。
```
Comment (csuzhangxc) [inline tools/dm/cluster-operations.md]: I think this step.4 can be removed now.

**Comment by dcalvin** [comment]
Thanks @csuzhangxc. PTAL again.

File: tools/dm/cluster-operations.md [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+        $ ansible-playbook -i hosts.ini create_users.yml -u root -k
+        ```
+
+        该步在 `172.16.10.74` 机器上创建了一个 `tidb` 用户，设置了 sudo 规则，并为中控机与该机器配置了 SSH 互信。
+
+2. 修改 `inventory.ini` 文件，创建新 DM-worker 实例 `dm_worker3`。
+
+    ```
+    [dm_worker_servers]
+    dm_worker1 source_id: "instance-1" ansible_host=172.16.10.72 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
```
Comment (dcalvin) [inline tools/dm/cluster-operations.md]: source_id: "instance-1" Is this the right format her? or we should write it like `source_id="instance-1"` @GregoryIan

**Comment by lilin90** [comment]
@GregoryIan @csuzhangxc PTAL again~ Updated the newly added source-id.

**Comment by lilin90** [comment]
@GregoryIan Please take a look and check whether the added "source-id"s are correct.

**Comment by dcalvin** [comment]
> Please pay special attention to links and guarantee your links can work well. You can check it by preview.

Fixed. Thanks. PTAL again @lilin90  Will PAY more attention in future.

**Comment by dcalvin** [comment]
> @GregoryIan Please take a look and check whether the added "source-id"s are correct.

Already confirmed this change with Ian before I updated. There is a similar instance to reference in[ DM deployment ](https://pingcap.com/docs-cn/tools/dm/deployment/).

File: tools/dm/cluster-operations.md [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+      要解决这个问题，请按照[手动处理 Sharding DDL Lock](/tools/dm/manually-handling-sharding-ddl-locks.md#场景二unlock-过程中部分-dm-worker-重启) 中描述的步骤操作。
+
+**总结**：尽量避免在 sharding DDL 同步过程中重启 DM-worker。
+
+#### DM-master 重启事项
+
+由 DM-master 维护的信息包括以下两种。重启 DM-master 不会持久化保存这些信息的相关数据。
+
+- 任务与 DM-master 之间的对应关系
```
Comment (GregoryIan) [inline tools/dm/cluster-operations.md]: `任务信息` 就够了

File: tools/dm/cluster-operations.md [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+1. 下载 DM 二进制文件。
+
+    1. 从 `downloads` 目录删除已有文件。 
+
+        ```
+        $ cd /home/tidb/dm-ansible
+        $ rm -rf downloads
+        ```
+
+    2. 用 Playbook 下载最新 DM 二进制文件。这会自动替换 `/home/tidb/dm-ansible/resource/bin/` 中已有文件。
```
Comment (GregoryIan) [inline tools/dm/cluster-operations.md]: 这个只能下载 latest verison 的吧

File: tools/dm/cluster-operations.md [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+        $ ansible-playbook -i hosts.ini create_users.yml -u root -k
+        ```
+
+        该步在 `172.16.10.74` 机器上创建了一个 `tidb` 用户，设置了 sudo 规则，并为中控机与该机器配置了 SSH 互信。
+
+2. 修改 `inventory.ini` 文件，创建新 DM-worker 实例 `dm_worker3`。
+
+    ```
+    [dm_worker_servers]
+    dm_worker1 source_id="instance-1" ansible_host=172.16.10.72 server_id=101 mysql_host=172.16.10.81 mysql_user=root mysql_password='VjX8cEeTX+qcvZ3bPaO4h0C80pe/1aU=' mysql_port=3306
```
Comment (GregoryIan) [inline tools/dm/cluster-operations.md]: 我建议和 deployment 文档保持一致，`source_id=mysql-replica-1`

**Comment by GregoryIan** [comment]
Rest LGTM

**Comment by lilin90** [comment]
@dcalvin Please help address the comments. Thanks!

File: tools/dm/cluster-operations.md (hunk 11 above)
Comment (dcalvin) [reply tools/dm/cluster-operations.md]: PTAL @liubo0127

**Comment by lilin90** [comment]
@dcalvin Would you please also help update the corresponding file in pingcap/docs based on Ian's comments? Thanks!

**Pull request merged**